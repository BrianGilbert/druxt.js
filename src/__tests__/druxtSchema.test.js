import { DruxtSchema } from '..'

jest.mock('axios')
jest.mock('simple-oauth2')

const baseURL = 'https://example.com'
const options = {}

let schema

describe('DruxtSchema', () => {
  beforeEach(() => {
    schema = new DruxtSchema(baseURL, options)
  })

  test('constructor', () => {
    // Throw error if 'baseURL' not provided.
    expect(() => { new DruxtSchema() }).toThrow('The \'baseURL\' parameter is required.')

    // Ensure class type.
    expect(new DruxtSchema(baseURL)).toBeInstanceOf(DruxtSchema)
  })

  test('get', async () => {
    const { schemas } = await schema.get()

    expect(Object.keys(schemas).length).toBe(4)

    expect(Object.values(schemas)[0]).toHaveProperty('id')
    expect(Object.values(schemas)[0]).toHaveProperty('resourceType')

    expect(Object.values(schemas)[0].config).toHaveProperty('entityType')
    expect(Object.values(schemas)[0].config).toHaveProperty('bundle')
    expect(Object.values(schemas)[0].config).toHaveProperty('mode')
    expect(Object.values(schemas)[0].config).toHaveProperty('schemaType')

    expect(schemas).toMatchSnapshot()
  })

  test('getResources', async () => {
    const resources = await schema.getResources('entity_view_display--entity_view_display')
    expect(Object.keys(resources).length).toBe(2)
    expect(Object.values(resources)[0]).toHaveProperty('attributes')
  })

  test('getSchema', async () => {
    // Ensure we don't get a filtered schema.
    const config = {
      entityType: 'node',
      bundle: 'page',
      filter: ['node--article--default--view']
    }
    expect(await schema.getSchema(config)).toBe(false)
  })

  test('getIndex', async () => {
    const index = await schema.getIndex()
    expect(Object.keys(index).length).toBe(12)
    expect(Object.values(index)[0]).toHaveProperty('href')
  })

  test('oauth2', async () => {
    schema = new DruxtSchema(baseURL, {
      auth: {
        type: 'oauth2',
        credentials: {
          clientId: 'clientId',
          clientSecret: 'clientSecret',
          username: 'username',
          password: 'password'
        }
      }
    })

    const callback = schema.oauth2()
    const request = await callback({ headers: {} })
    expect(request.headers).toHaveProperty('Authorization')
  })
})
