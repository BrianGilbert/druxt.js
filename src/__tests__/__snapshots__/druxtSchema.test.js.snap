// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtSchema get 1`] = `
Object {
  "articles--default--form": undefined,
  "articles--default--view": Object {
    "config": Object {
      "bundle": "article",
      "entityType": "node",
      "fields": Object {
        "body": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "body",
          "publicName": "body",
        },
        "changed": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "date_time",
            "settings": Object {
              "dateTimeFormat": "Y-m-d\\\\TH:i:sO",
            },
          },
          "fieldName": "changed",
          "publicName": "updatedAt",
        },
        "created": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "date_time",
            "settings": Object {
              "dateTimeFormat": "Y-m-d\\\\TH:i:sO",
            },
          },
          "fieldName": "created",
          "publicName": "createdAt",
        },
        "default_langcode": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "default_langcode",
          "publicName": "default_langcode",
        },
        "field_image": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "field_image",
          "publicName": "image",
        },
        "field_tags": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "field_tags",
          "publicName": "tags",
        },
        "langcode": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "langcode",
          "publicName": "langcode",
        },
        "nid": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "nid",
          "publicName": "internalId",
        },
        "path": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "nested",
            "settings": Object {
              "path": "alias",
            },
          },
          "fieldName": "path",
          "publicName": "path",
        },
        "promote": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "promote",
          "publicName": "isPromoted",
        },
        "revision_default": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_default",
          "publicName": "revision_default",
        },
        "revision_log": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_log",
          "publicName": "revision_log",
        },
        "revision_timestamp": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_timestamp",
          "publicName": "revision_timestamp",
        },
        "revision_translation_affected": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_translation_affected",
          "publicName": "revision_translation_affected",
        },
        "revision_uid": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_uid",
          "publicName": "revision_uid",
        },
        "status": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "status",
          "publicName": "isPublished",
        },
        "sticky": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "sticky",
          "publicName": "sticky",
        },
        "title": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "title",
          "publicName": "title",
        },
        "type": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "type",
          "publicName": "contentType",
        },
        "uid": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "uid",
          "publicName": "owner",
        },
        "uuid": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "uuid",
          "publicName": "uuid",
        },
        "vid": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "vid",
          "publicName": "vid",
        },
      },
      "mode": "default",
      "resourceType": "articles",
      "schemaType": "view",
    },
    "fields": Array [
      Object {
        "description": null,
        "id": "title",
        "label": Object {
          "position": "hidden",
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Object {
            "link_to_entity": false,
          },
        },
        "thirdPartySettings": Array [],
        "type": "string",
        "weight": -5,
      },
      Object {
        "description": null,
        "id": "field_tags",
        "label": Object {
          "position": "above",
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Object {
            "link": true,
          },
        },
        "thirdPartySettings": Array [],
        "type": "entity_reference_label",
        "weight": 0,
      },
      Object {
        "description": null,
        "id": "field_image",
        "label": Object {
          "position": "hidden",
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Object {
            "image_link": "",
            "responsive_image_style": "3_2_image",
          },
        },
        "thirdPartySettings": Array [],
        "type": "responsive_image",
        "weight": 1,
      },
      Object {
        "description": null,
        "id": "body",
        "label": Object {
          "position": "hidden",
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Array [],
        },
        "thirdPartySettings": Array [],
        "type": "text_default",
        "weight": 2,
      },
      Object {
        "description": null,
        "id": "links",
        "label": Object {
          "position": null,
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Array [],
        },
        "thirdPartySettings": Array [],
        "type": null,
        "weight": 3,
      },
    ],
    "groups": Array [],
    "id": "articles--default--view",
    "resourceType": "node--article",
  },
  "pages--default--form": undefined,
  "pages--default--view": Object {
    "config": Object {
      "bundle": "page",
      "entityType": "node",
      "fields": Object {
        "body": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "nested",
            "settings": Object {
              "path": "value",
            },
          },
          "fieldName": "body",
          "publicName": "body",
        },
        "changed": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "date_time",
            "settings": Object {
              "dateTimeFormat": "Y-m-d\\\\TH:i:sO",
            },
          },
          "fieldName": "changed",
          "publicName": "updatedAt",
        },
        "created": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "date_time",
            "settings": Object {
              "dateTimeFormat": "Y-m-d\\\\TH:i:sO",
            },
          },
          "fieldName": "created",
          "publicName": "createdAt",
        },
        "default_langcode": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "default_langcode",
          "publicName": "default_langcode",
        },
        "langcode": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "langcode",
          "publicName": "langcode",
        },
        "nid": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "nid",
          "publicName": "internalId",
        },
        "path": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "nested",
            "settings": Object {
              "path": "alias",
            },
          },
          "fieldName": "path",
          "publicName": "path",
        },
        "promote": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "promote",
          "publicName": "isPromoted",
        },
        "revision_log": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_log",
          "publicName": "revision_log",
        },
        "revision_timestamp": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_timestamp",
          "publicName": "revision_timestamp",
        },
        "revision_translation_affected": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_translation_affected",
          "publicName": "revision_translation_affected",
        },
        "revision_uid": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "revision_uid",
          "publicName": "revision_uid",
        },
        "status": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "status",
          "publicName": "isPublished",
        },
        "sticky": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "sticky",
          "publicName": "sticky",
        },
        "title": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "title",
          "publicName": "title",
        },
        "type": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "type",
          "publicName": "contentType",
        },
        "uid": Object {
          "disabled": false,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "uid",
          "publicName": "owner",
        },
        "uuid": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "uuid",
          "publicName": "uuid",
        },
        "vid": Object {
          "disabled": true,
          "enhancer": Object {
            "id": "",
          },
          "fieldName": "vid",
          "publicName": "vid",
        },
      },
      "mode": "default",
      "resourceType": "pages",
      "schemaType": "view",
    },
    "fields": Array [
      Object {
        "description": null,
        "id": "title",
        "label": Object {
          "position": "hidden",
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Object {
            "link_to_entity": false,
          },
        },
        "thirdPartySettings": Array [],
        "type": "string",
        "weight": -5,
      },
      Object {
        "description": null,
        "id": "body",
        "label": Object {
          "position": "hidden",
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Array [],
        },
        "thirdPartySettings": Array [],
        "type": "text_default",
        "weight": 100,
      },
      Object {
        "description": null,
        "id": "links",
        "label": Object {
          "position": null,
          "text": null,
        },
        "required": false,
        "settings": Object {
          "config": Object {},
          "display": Object {},
        },
        "thirdPartySettings": Object {},
        "type": null,
        "weight": 101,
      },
    ],
    "groups": Array [],
    "id": "pages--default--view",
    "resourceType": "node--page",
  },
}
`;
